### PR TITLE
KES fixes

### PR DESCRIPTION
**Pull request by mrBliss**
See the individual commit messages.

### CODE REVIEW

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs [hunk 1]
```diff
@@ -104,6 +108,28 @@ shelleyMaintainForgeState TPraosParams{..} (TPraosLeaderCredentials signKeyKES i
...
+-- <https://github.com/input-output-hk/cardano-ledger-specs/issues/1516>
+checkMaxKESEvolutions
+  :: forall c. TPraosCrypto c
+  => SL.ShelleyGenesis c
+  -> Either String ()
+checkMaxKESEvolutions genesis
+    | configuredMaxKESEvo > fromIntegral supportedMaxKESEvo
+    = throwError $ mconcat [
+          "sgMaxKESEvolutions greater than the supported evolutions"
+        , "by the KES algorithm: "
```
Comment (mrBliss) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node.hs:121]:
```suggestion
        , " by the KES algorithm: "
```

File: ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs [hunk 2]
```diff
@@ -109,8 +113,9 @@ prop_simple_real_tpraos_convergence TestSetup
     initialKESPeriod :: SL.KESPeriod
     initialKESPeriod = SL.KESPeriod 0
 
-    maxKESEvolution :: Word64
-    maxKESEvolution = 100 -- TODO
+    maxKESEvolutions :: Word64
```
Comment (mrBliss) [inline ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs:116]: Comment: a KES algorithm supports a particular max number of KES evolutions, but we can configure a potentially lower max for the ledger. Here, we just use the same value for both.

File: ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs [hunk 3]
```diff
@@ -127,7 +132,7 @@ prop_simple_real_tpraos_convergence TestSetup
           setupK
           setupD
           tpraosSlotLength
-          maxKESEvolution
```
Comment (mrBliss) [inline ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs:130]: Add the following comment to `mkGenesisConfig`: a KES algorithm supports a particular max number of KES evolutions, but we can configure a potentially lower max for the ledger.

File: ouroboros-consensus-shelley-test/src/Test/ThreadNet/Infra/Shelley.hs [hunk 4]
```diff
@@ -100,7 +100,7 @@ coreNodeKeys CoreNode{cnGenesisKey, cnDelegateKey, cnStakingKey, cnVRF, cnKES}
           , Gen.AllIssuerKeys
             { Gen.cold = mkDSIGNKeyPair cnDelegateKey
             , Gen.vrf  = mkVRFKeyPair cnVRF
-            , Gen.hot  = [(SL.KESPeriod 100, mkKESKeyPair cnKES)]
+            , Gen.hot  = [(SL.KESPeriod 0, mkKESKeyPair cnKES)]
             , Gen.hk   = SL.hashKey (SL.VKey $ deriveVerKeyDSIGN cnDelegateKey)
             }
```
Comment (mrBliss) [inline ouroboros-consensus-shelley-test/src/Test/ThreadNet/Infra/Shelley.hs:105]: Are KES or VRF keys even needed to generate transactions? (That's the only thing `coreNodeKeys` is used for.) Maybe this type is used for generating more things than transactions. Try replacing things with `error` to see whether they are actually used.

File: ouroboros-consensus-shelley-test/src/Test/ThreadNet/Infra/Shelley.hs [hunk 5]
```diff
@@ -89,28 +88,25 @@ data CoreNodeKeyInfo h = CoreNodeKeyInfo
...
           ( mkDSIGNKeyPair cnGenesisKey
           , Gen.AllIssuerKeys
             { Gen.cold = mkDSIGNKeyPair cnDelegateKey
-            , Gen.vrf  = mkVRFKeyPair cnVRF
-            , Gen.hot  = [(SL.KESPeriod 100, mkKESKeyPair cnKES)]
-            , Gen.hk   = SL.hashKey (SL.VKey $ deriveVerKeyDSIGN cnDelegateKey)
+              -- 'CoreNodeKeyInfo' is used for all sorts of generators, not
+              -- only transaction generators. To generate transactions we
+              -- don't need all these keys, hence the 'error's.
+            , Gen.vrf  = error "vrf used while generating transactions"
```
Comment (edsko) [inline ouroboros-consensus-shelley-test/src/Test/ThreadNet/Infra/Shelley.hs:101]: Nice

**Comment by mrBliss** [comment]
bors merge

**Pull request merged**